### PR TITLE
class library: improve suggestions in method not found error

### PR DESCRIPTION
**Pull request by telephon**
## Purpose and Motivation

The current error fixing suggestion can be misleading. Typically, the receiver is wrong, not the message. So when you call:

```
nil.flop
```

you get a suggestion that you mistyped `flop`, but it happens to be nil which doesn't understand flop:

```
ERROR: Message 'flop' not understood. Did you mean 'stop'?
RECEIVER:
   nil
ARGS:
```

This branch improves this situation.

If the receiver is nil, no suggestion is made - this seems like a reasonable thing not to do.

Otherwise, we post this:
```
"Did you mean '%'? Or should '%' have been called on another receiver?"
```

So that:
```supercollider
Set[1, 2, 3].flop 

// returns 

ERROR: Message 'flop' not understood by 'Set[ 1, 3, 2 ]'
Did you mean 'stop'? Or should 'flop' have been called on another receiver?
RECEIVER:
Instance of Set {    (0x11bbaff68, gc=78, fmt=00, flg=00, set=02)
  instance variables [2]
    array : instance of Array (0x11b705628, size=6, set=3)
    size : Integer 3
}
ARGS:
CALL STACK:
	DoesNotUnderstandError:reportError
		arg this = <instance of DoesNotUnderstandError>
	Nil:handleError
		arg 
[etc.]
```

This still isn't optimal in all cases, but catches most "misunderstandings".

## Types of changes

- general improvement

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] This PR is ready for review

### CODE REVIEW

**Comment by brianlheim** [comment]
thanks! but, i don't understand how turning off suggestions for nil solves the problem you raised.

**Comment by telephon** [comment]
I was just updating the text above, is it understandable now?

File: SCClassLibrary/Common/Core/Error.sc [hunk 1]
```diff
@@ -147,33 +147,36 @@ ShouldNotImplementError : MethodError {
...
+			}
 		}
 	}
 	errorString {
-		^"ERROR: Message '" ++ selector ++ "' not understood." ++ suggestion
+		^"ERROR: Message '%' not understood by '%'".format(selector, receiver)
 	}
 	reportError {
 		this.errorString.postln;
+		suggestion.postln;
```
Comment (brianlheim) [inline SCClassLibrary/Common/Core/Error.sc:179]: i really think this should be part of `errorString`, so it's printed both at the top and bottom of the log

File: SCClassLibrary/Common/Core/Error.sc [hunk 2]
```diff
@@ -147,33 +147,36 @@ ShouldNotImplementError : MethodError {
...
+			if(editDistances[minIndex] <= 3 and: { methodNames[minIndex].similarity(lowerCaseSelector) > 0 }) {
+				suggestedCorrection = methods[minIndex];
+				suggestion = "Did you mean '%'? Or should '%' have been called on another receiver?"
+				.format(suggestedCorrection.name, selector);
+			}
 		}
 	}
 	errorString {
-		^"ERROR: Message '" ++ selector ++ "' not understood." ++ suggestion
+		^"ERROR: Message '%' not understood by '%'".format(selector, receiver)
```
Comment (brianlheim) [inline SCClassLibrary/Common/Core/Error.sc:175]: i think this was better as it was. the receiver is already printed out in a more detailed form below this, and this could also lead to massive printouts if the receiver is a large string or collection.

File: SCClassLibrary/Common/Core/Error.sc [hunk 3]
```diff
@@ -147,33 +147,36 @@ ShouldNotImplementError : MethodError {
...
+			methodNames = methods.collect { |x| x.name.asString.toLower };
+			// we compare lower-case versions to prioritize capitalization mistakes
+			lowerCaseSelector = selector.asString.toLower;
+			editDistances = methodNames.collect(_.editDistance(lowerCaseSelector));
+			minIndex = editDistances.minIndex;
+			// Edit distance of 3 chosen arbitrarily; also, filter out completely dissimilar matches
+			// to avoid unhelpful suggestions.
+			if(editDistances[minIndex] <= 3 and: { methodNames[minIndex].similarity(lowerCaseSelector) > 0 }) {
+				suggestedCorrection = methods[minIndex];
+				suggestion = "Did you mean '%'? Or should '%' have been called on another receiver?"
```
Comment (brianlheim) [inline SCClassLibrary/Common/Core/Error.sc:169]: this feels a bit wordy to me. how's this: `"Perhaps you misspelled '%', or meant to call '%' on another receiver?"`

**Comment by telephon** [comment]
> makes sense now that i've had some time to think about it. i think your point is that messages not being understood by nil are far far more often due to a nil being passed where another object is intended, right?

Yes, it is the most common case, not only for nil, but also for many other cases like the one demonstrated above. I also find that an error system should rather give no suggestions than misleading ones.

File: SCClassLibrary/Common/Core/Error.sc (hunk 1 above)
Comment (telephon) [reply SCClassLibrary/Common/Core/Error.sc:179]: OK, for convenience this makes sense.

File: SCClassLibrary/Common/Core/Error.sc (hunk 2 above)
Comment (telephon) [reply SCClassLibrary/Common/Core/Error.sc:175]: ok!

**Comment by brianlheim** [comment]
> I also find that an error system should rather give no suggestions than misleading ones.

for the tools i use that implement this same kind of message, i find it immensely useful, and apparently the maintainers did too. i think we have to disagree here. :)

**Comment by telephon** [comment]
Don't get me wrong, I find these suggestions very useful. They only should be correct in a large majority of cases.

**Comment by telephon** [comment]
Better now? I've added a bit of space, this looks less cluttered to me now.

File: SCClassLibrary/Common/Core/Error.sc (hunk 3 above)
Comment (telephon) [reply SCClassLibrary/Common/Core/Error.sc]: much better.

File: SCClassLibrary/Common/Core/Error.sc [hunk 4]
```diff
@@ -147,42 +147,45 @@ ShouldNotImplementError : MethodError {
...
+		"\nRECEIVER:\n".post;
 		receiver.dump;
 		"ARGS:\n".post;
 		args.dumpAll;
 		this.errorPathString.post;
 		if(protectedBacktrace.notNil, { this.postProtectedBacktrace });
 		this.dumpBackTrace;
 		// this.adviceLink.postln;
-		"^^ The preceding error dump is for %\nRECEIVER: %\n\n\n".postf(this.errorString, receiver);
+		"\n^^ %\nRECEIVER: %\n\n\n".postf(this.errorString, receiver);
```
Comment (brianlheim) [inline SCClassLibrary/Common/Core/Error.sc:188]: please keep this as it was, the sections of the printout are very clearly labelled and extra vertical whitespace just leads to more scrolling.

Comment (telephon) [reply SCClassLibrary/Common/Core/Error.sc:188]: These are only 3 lines out of 40. Maybe give it a second look tomorrow, or we ask others what they think? These things are sometimes a matter of taste, in the end I'm fine with either of course. To me, it is an improvement.

Also I'd really like to get rid of the long "The preceding error dump is for" – it makes it harder to read, and the "^^" are clear indications upward.

Comment (brianlheim) [reply SCClassLibrary/Common/Core/Error.sc:188]: i don't want to solve formatting issues by adding extra blank lines in the middle of a printout; it creates a bad precedent, which i explained before [here](https://github.com/supercollider/supercollider/pull/4476#discussion_r419421204). i would be much more interested in removing unnecessary information, or changing the way the information is presented, but that would happen outside this PR.

Comment (telephon) [reply SCClassLibrary/Common/Core/Error.sc:188]: OK, I'm not sure I understand your problem, but fine with me! Let's keep this in mind though.
